lib/client/api.go

```diff
@@ -1060,13 +1072,24 @@ func (tc *TeleportClient) LoadKeyForCluster(clusterName string) error {
 func (tc *TeleportClient) LoadKeyForClusterWithReissue(ctx context.Context, clusterName string) error {
 	err := tc.LoadKeyForCluster(clusterName)
 	if err == nil {
+		// TODO(awly): is this always correct?
+		//
+		// Reset client AuthMethods, in case they were populated via relogin
+		// and would force us to use the root cert.
+		if len(tc.AuthMethods) > 0 {
+			tc.AuthMethods = nil
+		}
```

Comment:
I don't think this is needed at all. `tc.AuthMethods` (= `tc.Config.AuthMethods`) is never populated via relogin. It's always set explicitly alongside with `SkipLocalAuth = true`, i.e. `tc.AuthMethods` is never really involved in typical login scenarios. It's used either from `tsh` (when using an identity file) or the web terminal client.

Reply:
It's populated in https://github.com/gravitational/teleport/blob/fbae7ad508d0f30aacbfb2bf5a3878e1c448439b/lib/client/api.go#L472-L481

Here's what happens when relogin triggers in `RetryWithRelogin`:
- user logs in
- root cluster cert is issued
- root cluster cert is set in `tc.Config.AuthMethods`
- on next attempt, `ConnectToProxy` uses `tc.Config.AuthMethods[0]` first
- this cert is accepted by the leaf proxy, because CA validation succeeds
- the leaf proxy ends up routing the connection to `node@root` instead of `node@leaf` because of the cert cluster name extension
- the node connection fails because `leaf` can't forward requests to `root`

Because node connection happens after a successful `ConnectToProxy`, the client doesn't try any other `tc.authMethod`s.

Reply:
Right I didn't notice this occurrence.

This assignment to `tc.Config.AuthMethods` in `RetryWithRelogin` is just an optimization? Wouldn't it be easier to drop this assignment, and rely on the agent auth methods only...?

Reply:
You're right.
I was afraid to remove it, assuming something relies on the AuthMethods update.
But removing it doesn't seem to break anything on relogin (based on manual testing), so let's do that.

Worst case, we should catch any issues during the testplan.